core/state/statedb.go

```diff
@@ -802,28 +802,51 @@ func (s *StateDB) clearJournalAndRefund() {
 	s.validRevisions = s.validRevisions[:0] // Snapshots can be created without journal entires
 }
 
+// ModifiedAccount is an Account and it's Storage trie that has changed in the current block.
+type ModifiedAccount struct {
+	Account
+	Storage
+}
+
+// StateChanges are a map between an Account's address to it's ModifiedAccount.
+type StateChanges map[common.Address]ModifiedAccount
+
 // Commit writes the state to the underlying in-memory trie database.
-func (s *StateDB) Commit(deleteEmptyObjects bool) (common.Hash, error) {
+func (s *StateDB) Commit(deleteEmptyObjects bool) (common.Hash, StateChanges, error) {
 	if s.dbErr != nil {
-		return common.Hash{}, fmt.Errorf("commit aborted due to earlier error: %v", s.dbErr)
+		return common.Hash{}, nil, fmt.Errorf("commit aborted due to earlier error: %v", s.dbErr)
 	}
+
 	// Finalize any pending changes and merge everything into the tries
 	s.IntermediateRoot(deleteEmptyObjects)
 
+	stateChanges := make(StateChanges)
+
 	// Commit objects to the trie, measuring the elapsed time
 	for addr := range s.stateObjectsDirty {
+		modifiedAccount := ModifiedAccount{}
+
 		if obj := s.stateObjects[addr]; !obj.deleted {
 			// Write any contract code associated with the state object
 			if obj.code != nil && obj.dirtyCode {
 				s.db.TrieDB().InsertBlob(common.BytesToHash(obj.CodeHash()), obj.code)
 				obj.dirtyCode = false
 			}
+
+			// Add the account to the modifiedAccounts map
+			modifiedAccount = ModifiedAccount{Account: obj.data}
+			// Add the origin storage to the modifiedAccounts map before it's committed (and flushed from in-memory)
+			modifiedAccount.Storage = obj.originStorage
```

Comment:
I think the biggest lingering question I have is whether there's a way to eliminate diffs when a value is referenced but unchanged. Don't think that needs to block going upstream since the Geth maintainers probably have the best input about how to do that. That said I was looking at [this](https://github.com/ethereum/go-ethereum/blob/master/core/state/state_object.go#L319) and starting to wonder whether we might be able to make some progress by checking the `originStorage` against the `pendingStorage` 🤔

Reply:
Oh nice, that's a good call, I'll do a bit of digging into the `updateTrie` method you linked. I'm thinking that when I open a PR that I'll make a note about this piece and mention that we'd love some input.